src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -501,43 +530,63 @@ def expectation_from_circuit(self, circuit: "Circuit", nshots: int = 1000) -> fl
         coefficients = []
         Z_observables = []
         qubit_maps = []
-        for term in self.terms:
-            # store coefficient
-            coefficients.append(term.coefficient)
-            # Only care about non-I terms
-            non_identity_factors = [
-                factor for factor in term.factors if factor.name[0] != "I"
-            ]
-            # build diagonal observable
-            Z_observables.append(
-                SymbolicHamiltonian(
-                    prod(Z(factor.target_qubit) for factor in non_identity_factors),
-                    nqubits=circuit.nqubits,
-                    backend=self.backend,
+        # loop over the terms that can be diagonalized simultaneously
+        for terms in self.diagonal_terms:
+            # for each term that can be diagonalized simultaneously
+            # extract the coefficient, Z observable and qubit map
+            # the basis rotation, instead, will be the same
+            tmp_coeffs, tmp_obs, tmp_qmaps = [], [], []
+            measurements = {}
+            for term in terms:
+                # store coefficient
+                tmp_coeffs.append(term.coefficient)
+
+                # Only care about non-I terms
+                non_identity_factors = []
+                # prepare the measurement basis and append it to the circuit
+                for factor in term.factors:
+                    if factor.name[0] != "I":
+                        non_identity_factors.append(factor)
+                        q = factor.target_qubit
+                        if q not in measurements:
+                            measurements[q] = gates.M(q, basis=factor.gate.__class__)
+
+                # build diagonal observable
+                tmp_obs.append(
+                    SymbolicHamiltonian(
+                        prod(Z(factor.target_qubit) for factor in non_identity_factors),
+                        nqubits=circuit.nqubits,
+                        backend=self.backend,
+                    )
                 )
-            )
-            # Get the qubits we want to measure for each term
-            qubit_map = sorted(factor.target_qubit for factor in non_identity_factors)
-            # prepare the measurement basis and append it to the circuit
-            measurements = [
-                gates.M(factor.target_qubit, basis=factor.gate.__class__)
-                for factor in non_identity_factors
-            ]
+
+                # Get the qubits we want to measure for each term
+                tmp_qmaps.append(
+                    sorted(factor.target_qubit for factor in non_identity_factors)
+                )
+            coefficients.append(tmp_coeffs)
+            Z_observables.append(tmp_obs)
+            qubit_maps.append(tmp_qmaps)
```

Comment:
```suggestion
            qubit_maps.append(list(measurements.keys()))
```

I think you should be able to skip the whole `tmp_qmaps` thing with this.